src/UserSpaceInstrumentation/Trampoline.cpp

```diff
@@ -77,6 +77,60 @@ constexpr uint64_t kOffsetOfFunctionIdInCallToEntryPayload = 178;
   return result;
 }
 
+// Check if somewhere in the code of `function` there is a (conditional) jump back to the first five
+// bytes of the function (which we intend to overwrite with a jump into the trampoline). If so we
+// must not instrument the function. Note that the entire function is not necessarily available
+// here; we'll just disassemble and check whatever we have.
+// This is merely a heuristic. There can be other jumps either further down in the function or in
+// different places in the same translation unit that target the first five bytes of a function.
+// However analysing existing code shows that many of the problematic jumps are in small functions
+// that are written in assembly. These are detected by the function below.
+bool CheckForJumpIntoFirstFiveBytes(uint64_t function_address, const std::vector<uint8_t>& function,
+                                    csh capstone_handle) {
+  cs_insn* instruction = cs_malloc(capstone_handle);
+  ORBIT_FAIL_IF(instruction == nullptr, "Failed to allocate memory for capstone disassembler.");
+  orbit_base::unique_resource scope_exit{instruction,
+                                         [](cs_insn* instruction) { cs_free(instruction, 1); }};
+
+  const uint8_t* code_pointer = function.data();
+  size_t code_size = function.size();
+  uint64_t disassemble_address = function_address;
+
+  // Disassemble until we processed the first kBytesToDisassemble bytes or we run out of
+  // instructions in this function.
+  while (cs_disasm_iter(capstone_handle, &code_pointer, &code_size, &disassemble_address,
+                        instruction)) {
+    const uint64_t original_instruction_address = disassemble_address - instruction->size;
+    if ((instruction->detail->x86.opcode[0] == 0xeb) ||
+        ((instruction->detail->x86.opcode[0] & 0xf0) == 0x70)) {
+      // 0xeb is an uncoditional jump to an 8 bit immediate offset.
+      // 0x7? are conditional jumps to an 8 bit immediate offset.
+      const int8_t immediate = *absl::bit_cast<int8_t*>(
+          instruction->bytes + instruction->detail->x86.encoding.imm_offset);
+      const uint64_t jump_target_address =
+          original_instruction_address + instruction->size + immediate;
```

Comment:
Do you need `original_instruction_address`? Isn't `original_instruction_address + instruction->size` just `disassemble_address`?